pkg/config/manager.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"crypto/sha1"
+	"io/ioutil"
+	"time"
+
+	"sync"
+
+	"github.com/golang/glog"
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/pkg/expr"
+)
+
+// ChangeListener listens for config change notifications.
+type ChangeListener interface {
+	ConfigChange(cfg *Runtime)
+}
+
+// Manager represents the config Manager.
+// It is responsible for fetching and receiving configuration changes.
+// It applied validated changes to the registered config change listeners.
+// api.Handler listens for config changes.
+type Manager struct {
+	ManagerArgs
+
+	cl      []ChangeListener
+	closing chan bool
+	scSha   [sha1.Size]byte
+	gcSha   [sha1.Size]byte
+
+	lock      *sync.RWMutex
+	lastError error
+}
+
+// ManagerArgs are constructor args for a manager
+type ManagerArgs struct {
```

Comment:
Same as before, I think the struct-as-args is an anti-pattern. Where I've seen struct-as-args used the struct represents options that affect the behavior of the thing being constructed, and they're typically optional. In this case all the arguments are required (except loop delay, I suppose), and you even embed the Args struct directly into the manager. You also don't get any encapsulation since all of these fields are public to enable construction outside of the package; since the Args struct is embedded in the Manager I can change the values arbitrarily after construction. Even though 6 args to the NewManager func isn't great, I think it's better than the struct version.

Alternatively, do what [gRPC does, have a private options struct](https://github.com/grpc/grpc-go/blob/master/server.go#L107) (or, in this case, args struct) and provide functions to operate on them. But, since all of these args are required that just forces more boilerplate everywhere.